Add standard Linux paths to libpq.def

matobet opened the pull request:
Add `/usr/include` and `/usr/lib64` to compiler and linker options respectively. 

This should cover majority of x64 linux systems.

matobet committed:
Add standard Linux paths to libpq.def

Add `/usr/include` and `/usr/lib64` to compiler and linker options
respectively. This should cover majority of x64 linux systems.

hfhbd commented:
Can you please add a reason why do you need it? These paths are for development only, afaik.

matobet commented:
@hfhbd Without this change the project doesn't build under Linux (Fedora 38, but I assume more Linux systems will also suffer from this).

matobet committed:
libpq.def: linux_x64 settings in separate override

matobet commented:
Not sure why the linuxArm64 suddenly started failing (I have trouble cross-compiling linux_x64 to linux_arm64 on my system both before as well as after my change...) but perhaps moving the settings into platform-specific `.linux_x64` sections will fix the CI failure...

hfhbd commented:
I think the failures are caused to the fact the local `usr/include` contains unrelated header files but the filter is set to `*` allowing all headers.

matobet committed:
libpq.def: Narrow headerFilter

matobet commented:
Another shot with a more narrow `headerFilter`

Myshkouski commented:
> Add `/usr/include` and `/usr/lib64` to compiler and linker options respectively.

+1  and here the reasons why.

> Can you please add a reason why do you need it? These paths are for development only, afaik.

These paths are also for both development of dependent projects and runtime as well because `libpq-dev` provides [dynamic (shared) libraries](https://medium.com/@4318_26766/creating-and-using-dynamic-libraries-in-c-language-912b078b7c52).

You might build a [static native library](https://kotlinlang.org/docs/multiplatform-build-native-binaries.html#declare-binaries) instead, but your binary wouldn't be published with generated JARs, so other developers cannot use it as a dependency from Maven.

hfhbd commented:
@matobet I added some paths, do you need more?

matobet commented:
@hfhbd it seems on Fedora I still need to add `-I/usr/include` since the `libpq-devel` package installs the files there
```
/usr/bin/pg_config
/usr/include/libpq
/usr/include/libpq-events.h
/usr/include/libpq-fe.h
/usr/include/libpq/libpq-fs.h
/usr/include/pg_config-x86_64.h
/usr/include/pg_config.h
/usr/include/pg_config_ext-x86_64.h
/usr/include/pg_config_ext.h
/usr/include/pg_config_manual.h
/usr/include/pg_config_os.h
/usr/include/pgsql
/usr/include/pgsql/internal
/usr/include/pgsql/internal/c.h
/usr/include/pgsql/internal/fe-auth-sasl.h
/usr/include/pgsql/internal/libpq
/usr/include/pgsql/internal/libpq-int.h
/usr/include/pgsql/internal/libpq/pqcomm.h
/usr/include/pgsql/internal/port.h
/usr/include/pgsql/internal/postgres_fe.h
/usr/include/pgsql/internal/pqexpbuffer.h
/usr/include/postgres_ext.h
/usr/lib/.build-id
/usr/lib/.build-id/a8
/usr/lib/.build-id/a8/d9850aecf80d6f1207645fb9b1a67e0050cd49
/usr/lib64/libpq.so
/usr/lib64/pkgconfig/libpq.pc
```
(without the `-I/usr/include` I get the `fatal error: 'libpq-fe.h' file not found` error when executing the `:postgres-native-sqldelight-driver:cinteropLibpqLinuxX64` task)

hfhbd commented:
Okay, can you use update the PR? Thanks.

matobet committed:
Merge remote-tracking branch 'upstream/main'

matobet commented:
The tricky part here is to come up with such paths that make the project build *both* on fedora && on ubuntu (CI)...

matobet committed:
Merge remote-tracking branch 'upstream/main'

matobet committed:
libpq.def: use -idirafter for /usr/include, add aarch64 lib path

-I/usr/include lets host glibc headers shadow the Kotlin/Native
target sysroot, breaking cross-target cinterop (e.g. linuxArm64 on
an x64 host fails with 'gnu/stubs-32.h' not found). -idirafter keeps
/usr/include as a last-resort search dir so system-installed libpq
headers are still found, while the sysroot wins for system includes.

Also add -L/usr/lib/aarch64-linux-gnu for arm64 Debian/Ubuntu.
